src/policies/storage.ts

```diff
@@ -0,0 +1,87 @@
+import { a } from "pepr";
+
+import { When, containers, volumes } from "./common";
+import { exemptHostPathWrite, exemptVolumeType } from "./exemptions/storage";
+
+/**
+ * Restrict Volume Types for Pods
+ *
+ * Volume types, beyond the core set, should be restricted to limit exposure to potential vulnerabilities
+ * in Container Storage Interface (CSI) drivers.  In addition, HostPath volumes should not be allowed
+ * because host directories could be exploited to access shared data or escalate privileges.  This policy
+ * restricts use of volume types to the allowed list.
+ *
+ * @related https://repo1.dso.mil/big-bang/product/packages/kyverno-policies/-/blob/main/chart/templates/restrict-volume-types.yaml
+ */
+When(a.Pod)
+  .IsCreatedOrUpdated()
+  .Validate(request => {
+    // List of allowed volume types
+    const allowedVolumeTypes = [
+      "configMap",
+      "csi",
+      "downwardAPI",
+      "emptyDir",
+      "ephemeral",
+      "persistentVolumeClaim",
+      "projected",
+      "secret",
+    ];
+
+    if (exemptVolumeType(request)) {
+      return request.Approve();
+    }
+
+    // Check all volumes in the pod spec, if any
+    for (const volume of volumes(request)) {
+      // Get the volume type, which will be the only key in the volume object other than "name"
+      const volumeType = Object.keys(volume).find(key => key !== "name") || "unknown";
+
+      // If the volume type is not in the allowed list, deny the request
+      if (!allowedVolumeTypes.includes(volumeType)) {
+        return request.Deny(`Volume ${volume.name} has an invalid volume type of '${volumeType}.`);
```

Comment:
```suggestion
        return request.Deny(`Volume ${volume.name} has a disallowed volume type of '${volumeType}'.`);
```
nit: also maybe should log allowed types?